plugins/inputs/cloud_pubsub/pubsub.go

```diff
@@ -119,13 +127,46 @@ func (ps *PubSub) Stop() {
 	ps.wg.Wait()
 }
 
-func (ps *PubSub) subReceive(cctx context.Context) {
+// startReceiver is called within a goroutine and manages keeping a
+// subscription.Receive() up and running while the plugin has not been stopped.
+func (ps *PubSub) receiveWithRetry() {
+	err := ps.startReceiver()
+
+	for err != nil && ps.ctx.Err() == nil {
+		log.Printf("E! Receiver for subscription %s exited with error: %v", ps.sub.ID(), err)
+
+		delay := defaultRetryDelaySeconds
+		if ps.RetryReceiveDelaySeconds > 0 {
+			delay = ps.RetryReceiveDelaySeconds
+		}
+
+		log.Printf("I! Waiting %d seconds before attempting to restart receiver...", delay)
+		time.Sleep(time.Duration(delay) * time.Second)
+
+		err = ps.startReceiver()
+	}
+}
+
+func (ps *PubSub) startReceiver() error {
+	cctx, ccancel := context.WithCancel(ps.ctx)
+
+	log.Printf("I! Starting receiver for subscription %s...", ps.sub.ID())
+	return ps.receiveOnSub(cctx, ccancel)
+}
+
+func (ps *PubSub) receiveOnSub(cctx context.Context, ccancel context.CancelFunc) error {
 	err := ps.sub.Receive(cctx, func(ctx context.Context, msg message) {
 		if err := ps.onMessage(ctx, msg); err != nil {
 			ps.acc.AddError(fmt.Errorf("unable to add message from subscription %s: %v", ps.sub.ID(), err))
 		}
 	})
-	ps.acc.AddError(fmt.Errorf("receiver for subscription %s exited: %v", ps.sub.ID(), err))
+	if err != nil {
+		ps.acc.AddError(fmt.Errorf("receiver for subscription %s exited: %v", ps.sub.ID(), err))
+	} else {
+		log.Printf("I! subscription pull ended (no error, most likely stopped)")
+	}
+	ccancel()
```

Comment:
I'm confused here, can you explain what this cancel function is doing?  It seems like the operation is over at this point?

Reply:
I'm just making sure the child context I create per subscription receive is canceled. I believe this needs to be done to guarantee that some message handling is finished at the end of the receive (i.e. if receive fails), and it doesn't hurt even if the Google lib already does it. This was previously being done on the top-level cancel during Stop().